### PR TITLE
darwin/firewall: Fixes for alf_exceptions, make alf_services an alias for sharing_preferences

### PR DESCRIPTION
**Pull request by woodruffw**
This adds the `pathFromNestedPlistAliasData` function to handle an edge case in parsing some of the `alias` fields in the ALF plist, and turns `alf_services` into an alias for `sharing_preferences` (the original backing plist was vestigial and didn't accurately represent the state of the application firewall).

### CODE REVIEW

File: specs/darwin/alf_services.table [hunk 1]
```diff
@@ -1,8 +1,15 @@
 table_name("alf_services")
 description("OS X application layer firewall (Firewall) services.")
 schema([
-    Column("service", TEXT, "Firewalled service name"),
-    Column("process", TEXT, "Process name"),
-    Column("state", INTEGER, "Firewall service state"),
```
Comment (akindyakov) [inline specs/darwin/alf_services.table]: I'd suggest to keep this columns, but mark them as deprecated until the next major release. It's not fair to change interface significantly without major version update. What do you think?

Comment (woodruffw) [reply specs/darwin/alf_services.table]: Sounds good! Should I stub them out with default values, or leave them in as-is?

Comment (akindyakov) [reply specs/darwin/alf_services.table]: I didn't think about it, but there are some folks who have queries based on that columns. Why do we need to remove them? Does table only with numbers in columns make some sense?

Comment (woodruffw) [reply specs/darwin/alf_services.table]: > Why do we need to remove them?

They don't reflect the actual firewall state. Here's what `alf_services` looks like when I have Apple Events and Printer Sharing **enabled**:

```
osquery> select * from alf_services;
+-----------------------+---------------------+-------+
| service               | process             | state |
+-----------------------+---------------------+-------+
| Personal Web Sharing  | httpd               | 0     |
| Printer Sharing       | cupsd               | 0     |
| Remote Apple Events   | AEServer            | 0     |
| FTP Access            | ftpd                | 0     |
| Personal File Sharing | AppleFileServer     | 0     |
| Remote Login - SSH    | sshd-keygen-wrapper | 0     |
| Samba Sharing         | smbd                | 0     |
| Apple Remote Desktop  | AppleVNCServer      | 0     |
| ODSAgent              | ODSAgent            | 0     |
+-----------------------+---------------------+-------+
```

what I **expect** is this:

```
osquery> select * from alf_services;
+-----------------------+---------------------+-------+
| service               | process             | state |
+-----------------------+---------------------+-------+
| Personal Web Sharing  | httpd               | 0     |
| Printer Sharing       | cupsd               | 1     |
| Remote Apple Events   | AEServer            | 1     |
| FTP Access            | ftpd                | 0     |
| Personal File Sharing | AppleFileServer     | 0     |
| Remote Login - SSH    | sshd-keygen-wrapper | 0     |
| Samba Sharing         | smbd                | 0     |
| Apple Remote Desktop  | AppleVNCServer      | 0     |
| ODSAgent              | ODSAgent            | 0     |
+-----------------------+---------------------+-------+
```

Essentially, the underlying store (`firewall` in `com.apple.alf.plist`) is no longer used by macOS.  `sharing_preferences` is the only table (AFAIK) that accurately reflects service state.

**Comment by woodruffw** [comment]
Alright, I've added integration tests for `alf_services` and `alf_exceptions`.

**Comment by woodruffw** [comment]
Anything else I can do to help get this merged, @akindyakov?

**Comment by woodruffw** [comment]
Polite ping on this. Is there anything blocking another review/merge?

File: osquery/tables/system/darwin/firewall.cpp [hunk 2]
```diff
@@ -94,12 +95,16 @@ QueryData parseALFExceptionsTree(const pt::ptree& tree) {
...
       std::string path;
       auto alias_data = it.second.get<std::string>("alias", "");
+      auto status = pathFromNestedPlistAliasData(alias_data, path);
 
-      if (pathFromPlistAliasData(alias_data, path).ok()) {
-        r["path"] = path;
-        r["state"] = INTEGER(it.second.get("state", -1));
-        results.push_back(r);
+      if (!status.ok()) {
+        TLOG << status.getMessage();
```
Comment (theopolis) [inline osquery/tables/system/darwin/firewall.cpp]: I think we should add some context to the error message: `"Could not parse nested plist for applications: " << status.getMessage()`. Or something more accurate compared to my example. ;)

File: osquery/tables/system/darwin/firewall.h [hunk 3]
```diff
@@ -29,10 +29,8 @@ osquery::QueryData parseALFExceptionsTree(const pt::ptree& tree);
 // parseALFExplicitAuthsTree parses out the "explicitauth" key
 osquery::QueryData parseALFExplicitAuthsTree(const pt::ptree& tree);
 
-// Given a property tree of the parsed content of com.apple.alf.plist,
-// parseALFServicesTree parses out the services which exist under the
-// "firewall" key
-osquery::QueryData parseALFServicesTree(const pt::ptree& tree);
+// An alias for the sharing_preferences table
+osquery::QueryData parseALFServicesTree();
```
Comment (theopolis) [inline osquery/tables/system/darwin/firewall.h]: We can remove this specific code and define the alias in the `sharing_preferences` specfile. 

```py
table_name("sharing_preferences", aliases=[THIS_TABLE_NAME])
```

File: osquery/utils/darwin/plist.mm [hunk 4]
```diff
@@ -277,4 +283,22 @@ Status pathFromPlistAliasData(const std::string& data, std::string& result) {
   CFRelease(bookmark);
   return status;
 }
+
+/// Parse a nested (base-64 encoded) plist's alias data for its path
+/// Observed experimentally with /Library/Preferences/com.apple.alf.plist
+Status pathFromNestedPlistAliasData(const std::string& data,
```
Comment (theopolis) [inline osquery/utils/darwin/plist.mm]: Do you mind uploading test data into `./tests` and making a unit test for this new parsing. 

I can +1 to the observation, it's the same format on a 10.14.6 test device.

Comment (woodruffw) [reply osquery/utils/darwin/plist.mm]: Yeah, let me see if I can whip up a sanitized version of a local sample.

**Comment by theopolis** [comment]
Looking good! Last thing needed is the unit test.

**Comment by woodruffw** [comment]
Looks like the CI is failing on Linux due to disk exhaustion?

```
BUILT 975/1073 JOBS 21.0s //osquery/events/tests:events_database_tests#binary
[2019-10-02 16:22:42.356][error][command:73980f72-6344-4cb4-88dc-4131143599e9][tid:27][com.facebook.buck.artifact_cache.ArtifactUploader] When creating artifact archive for //osquery/events/tests:inotify_tests#binary containing: 
 buck-out/release/bin/osquery/events/tests/.inotify_tests#binary/metadata/artifact
buck-out/release/bin/osquery/events/tests/.inotify_tests#binary/metadata/artifact/OUTPUT_HASH
buck-out/release/bin/osquery/events/tests/.inotify_tests#binary/metadata/artifact/OUTPUT_SIZE
buck-out/release/bin/osquery/events/tests/.inotify_tests#binary/metadata/artifact/RECORDED_PATHS
buck-out/release/bin/osquery/events/tests/.inotify_tests#binary/metadata/artifact/RECORDED_PATH_HASHES
buck-out/release/gen/osquery/events/tests/inotify_tests
buck-out/release/gen/osquery/events/tests/inotify_tests#binary.: java.io.IOException: No space left on device

[2019-10-02 16:22:42.356][error][command:73980f72-6344-4cb4-88dc-4131143599e9][tid:28][com.facebook.buck.artifact_cache.ArtifactUploader] When creating artifact archive for //osquery/events/tests:fsevents_tests#binary containing: 
 buck-out/release/bin/osquery/events/tests/.fsevents_tests#binary/metadata/artifact
buck-out/release/bin/osquery/events/tests/.fsevents_tests#binary/metadata/artifact/OUTPUT_HASH
buck-out/release/bin/osquery/events/tests/.fsevents_tests#binary/metadata/artifact/OUTPUT_SIZE
buck-out/release/bin/osquery/events/tests/.fsevents_tests#binary/metadata/artifact/RECORDED_PATHS
buck-out/release/bin/osquery/events/tests/.fsevents_tests#binary/metadata/artifact/RECORDED_PATH_HASHES
buck-out/release/gen/osquery/events/tests/fsevents_tests
buck-out/release/gen/osquery/events/tests/fsevents_tests#binary.: java.io.IOException: No space left on device

BUILT 976/1073 JOBS 16.6s //osquery/events/tests:inotify_tests#binary
BUILT 977/1073 JOBS 20.6s //osquery/events/tests:fsevents_tests#binary
double free or corruption (out)
##[error]Bash exited with code '134'.
##[section]Finishing: Run tests
```

**Comment by theopolis** [comment]
@mike-myers-tob or @Smjert, see above about disk exhaustion. Is this something we have control over within Azure?

**Pull request merged**